// merlin-node/lib/test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const _1 = require(".");
const merlin = new _1.Merlin({
    apiKey: 'OPENAI_API_KEY',
    merlinConfig: {
        apiKey: 'Mx0U7w1b3t9wTppXa1My',
    },
});
const test = async () => {
    const response = await merlin.chat.completions.create({
        messages: [
            {
                role: 'assistant',
                content: 'You are a helpful assistant.',
            },
            {
                role: 'user',
                content: 'Can you tell me about mahatma gandhi.',
            },
        ],
        model: 'codellama/CodeLlama-34b-Instruct-hf',
        stream: true,
    });
    const [stream1, stream2] = response.tee();
    const stream = response.toReadableStream();
    for await (const chunk of stream1) {
        console.log(JSON.stringify(chunk, null, 2));
    }
};
const test2 = async () => {
    const response = await merlin.images.generate({
        prompt: 'This is a picture of a dog',
        // model: 'dall-e-2',
        model: 'Dreamshape v7',
        size: '256x256',
        response_format: 'b64_json',
    });
    console.log(JSON.stringify(response, null, 2));
};
void test();
